homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,205 @@
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
+        vol.Required(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    },
+    mqtt.validate_device_has_at_least_one_identifier,
+)
+
+DEVICE_TRIGGERS = "mqtt_device_triggers"
+
+
+@attr.s(slots=True, frozen=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    type = attr.ib(type=str)
+    subtype = attr.ib(type=str)
+    topic = attr.ib(type=str)
+    payload = attr.ib(type=str)
+    qos = attr.ib(type=int)
+    discovery_hash = attr.ib(type=str)
+
+
+def _update_trigger(hass, discovery_hash, config):
+    for device_id, triggers in hass.data[DEVICE_TRIGGERS].items():
+        if discovery_hash in triggers:
+            triggers[discovery_hash] = Trigger(
+                device_id=device_id,
+                type=config[CONF_TYPE],
+                subtype=config[CONF_SUBTYPE],
+                topic=config[CONF_TOPIC],
+                payload=config[CONF_PAYLOAD],
+                qos=config[CONF_QOS],
+                discovery_hash=discovery_hash,
+            )
+
+
+def _remove_trigger(hass, discovery_hash):
+    for _, triggers in hass.data[DEVICE_TRIGGERS].items():
+        triggers.pop(discovery_hash)
+
+
+async def async_setup_trigger(
+    hass, config, async_add_entities, config_entry, discovery_hash
+):
+    """Set up the MQTT device trigger."""
+    config = PLATFORM_SCHEMA(config)
+    remove_signal = None
+
+    @callback
+    def discovery_callback(payload):
+        """Handle discovery update."""
+        _LOGGER.info(
+            "Got update for trigger with hash: %s '%s'", discovery_hash, payload
+        )
+        if not payload:
+            # Empty payload: Remove trigger
+            _LOGGER.info("Removing trigger: %s", discovery_hash)
+            _remove_trigger(hass, discovery_hash)
+            clear_discovery_hash(hass, discovery_hash)
+            remove_signal()
+        else:
+            # Non-empty payload: Update trigger
+            _LOGGER.info("Updating trigger: %s", discovery_hash)
+            payload.pop(ATTR_DISCOVERY_HASH)
+            config = PLATFORM_SCHEMA(payload)
+            _update_trigger(hass, discovery_hash, config)
+
+    remove_signal = async_dispatcher_connect(
+        hass, MQTT_DISCOVERY_UPDATED.format(discovery_hash), discovery_callback
+    )
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    config_entry_id = config_entry.entry_id
+    device_info = mqtt.device_info_from_config(config[CONF_DEVICE])
+
+    if config_entry_id is not None and device_info is not None:
+        device_info["config_entry_id"] = config_entry_id
+        device_registry.async_get_or_create(**device_info)
```

Comment:
Why this? The entry needs to be registered or else the user couldn't have configured this to begin with ?

Reply:
There's no separate MQTT discovery message for defining devices, it's done by including a `device` key in the config struct.